tools/tpm2_createek.c

```diff
@@ -244,15 +367,27 @@ static bool on_option(char key, char *value) {
     case 'p':
         ctx.auth_ek.auth_str = value;
         break;
-    case 'G': {
-        TPMI_ALG_PUBLIC type = tpm2_alg_util_from_optarg(value,
-                tpm2_alg_util_flags_base);
-        if (type == TPM2_ALG_ERROR) {
-            LOG_ERR("Invalid key algorithm, got \"%s\"", value);
-            return false;
+    case 'G':
+       if (!strncmp("ecc", value, 3)) {
```

Comment:
Wouldn't it just be simpler to delay handling of this and set a field to string "rsa2048:null:aes128cfb." and override it here?

Then in the main flow you can just set up the object with a call to this and the _ext variant supports raw integer values IIRC.

Reply:
I don't understand this approach about delayed handling of the parameters. probably because I'm not too familiar with your parsers. `tpm2_alg_util_handle_ext_alg` is not capable of handling raw integer values.

Reply:
Apparently I went to edit my comment and never hit submit and lost that tab.... anyways sorry for the confusion here.

This should really be around using tpm2_alg_util_public_init(). Where you can specify all the parts and get back a public as expected. Right now, the code manually populates all the fields, when this function will do the heavy lifting for you and tons of the code in the tool just go away.

Reply:
The problem is that ` tpm2_alg_util_public_init(char *alg_details, char *name_halg, ...)` requires `name_halg` as an input argument because otherwise `nameAlg` would currently be set to the default `TPM2_ALG_SHA256`   in the populated `public` structure . Also I would need to know the symmetric encryption algorithm in advance in order to append it to the  `alg_details` input string. Thus I would either have to preparse `alg_details` first to determine all the input parameters needed by  `tpm2_alg_util_public_init()` or add a switch statement very similar to the existing `set_key_algorithm()` after the call so that not much is gained by calling  `tpm2_alg_util_public_init()` first.

The only way of saving code lines would be for `tpm2_alg_util_public_init()` to set reasonable defaults for nameAlg and symmetric encryption depending on the value of the key algorithm type.

Reply:
I think we can map the expected input string, like `ecc384` to what matches the templates:
1. the full specifier that works for `tpm2_alg_util_public_init`
2. The nameAlg per the template

So the vast majority of the code goes away and now you have something like:
```C
typedef struct alg_map alg_map;
struct alg_map {
   const char *input;
   const char *alg;
   const char *namealg;
   uint8_t *policy;
   const char *attributes;
};

static const uint8_t policy384[] = {
0xB2, 0x6E, 0x7D, 0x28, 0xD1, 0x1A,
0x50, 0xBC, 0x53, 0xD8, 0x82, 0xBC,
0xF5, 0xFD, 0x3A, 0x1A, 0x07, 0x41,
0x48, 0xBB, 0x35, 0xD3, 0xB4, 0xE4,
0xCB, 0x1C, 0x0A, 0xD9, 0xBD, 0xE4,
0x19, 0xCA, 0xCB, 0x47, 0xBA, 0x09,
0x69, 0x96, 0x46, 0x15, 0x0F, 0x9F,
0xC0, 0x00, 0xF3, 0xF8, 0x0E, 0x12
};

/*
 * See: https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_EKCredentialProfile_v2p4_r3.pdf
 */
alg_map algs[] = {
// from B.4.6 Template H-3: ECC NIST P384 (Storage) (I don;t know how 256 works since storage parents are required to 
// be aes128cfb for duplication IIUC).
{ "ecc384", "ecc384:aes256cfb", "sha384", policy384, "fixedtpm|fixedparent|sensitivedataorigin|userwithauth|adminwithpolicy|restricted|decrypt" },
...
};

alg_map m = lookup(userinput);
if (!m) {
 do_invalid_error();
}

TPM2B_PUBLIC public;
tpm2_alg_util_public_init(m->alg, m->namealg, m->attrs, NULL, def_attrs, &public);

/* set policy hash here */
memset(...);

....

create_primary()
```